### PR TITLE
#513 Prop change event

### PR DESCRIPTION
**Pull request by jpnelson**

### CODE REVIEW

File: src/lifecycle/properties-init.js [hunk 1]
```diff
@@ -17,6 +19,21 @@ function getLinkedAttribute (name, attr) {
   return attr === true ? dashCase(name) : attr;
 }
 
+function triggerPropertyChangedEvent (elem, {propertyName, eventName, oldValue, newValue}) {
+  let propertyChangeEvent = createCustomEvent(eventName, {
```
Comment (treshugart) [inline src/lifecycle/properties-init.js]: `const`

File: src/lifecycle/properties-init.js [hunk 2]
```diff
@@ -17,6 +19,21 @@ function getLinkedAttribute (name, attr) {
...
+    bubbles: false,
+    cancelable: true,
+    detail: {
+      name: propertyName,
```
Comment (treshugart) [inline src/lifecycle/properties-init.js]: You can just just do:

```js
{ name, oldValue, newValue }
```

Where `name` is `propertyName`.

File: README.md [hunk 3]
```diff
@@ -461,6 +461,22 @@ The parameters passed to the function are:
...
+##### `event`
+
+An event name to trigger whenever the property changes. This event is cancelable, and does not bubble.
+
+```js
+skate('my-component', {
+  properties: {
+    myProp: {
+      event: 'my-prop-changed'
+  }
```
Comment (treshugart) [inline README.md]: Needs closing `}` for the `myProp` definition.

File: src/api/emit.js [hunk 4]
```diff
@@ -1,25 +1,8 @@
 import createElement from '../native/create-element';
-import createEvent from '../native/create-event';
+import createCustomEvent from '../util/create-custom-event';
```
Comment (treshugart) [inline src/api/emit.js]: Is there any reasoning for separating this out and not just using `emit()` directly? I guess you don't need all the behaviour that emit provides. More curious that anything else.

File: src/lifecycle/properties-init.js [hunk 5]
```diff
@@ -17,6 +19,21 @@ function getLinkedAttribute (name, attr) {
...
+      oldValue: oldValue,
+      newValue: newValue
+    }
+  });
+
+  const cancelled = !dispatch(elem, propertyChangeEvent);
```
Comment (treshugart) [inline src/lifecycle/properties-init.js]: Could just return here.

File: test/unit/api/properties.js [hunk 6]
```diff
@@ -20,6 +20,70 @@ function testTypeValues (type, values) {
 }
 
 describe('api/properties', function () {
+  describe.only('property change event', function () {
```
Comment (treshugart) [inline test/unit/api/properties.js]: `.only` ;)

File: test/unit/api/properties.js [hunk 7]
```diff
@@ -20,6 +20,70 @@ function testTypeValues (type, values) {
...
+
+    it('contains property name, and change details', function () {
+      let elem = create(assign({event: 'propertychange'}, properties.boolean()));
+      let event = null;
+      elem.test = false;
+
+      elem.addEventListener('propertychange', (e) => event = e);
+      elem.test = true;
+
+      expect(event).to.not.equal(null);
```
Comment (treshugart) [inline test/unit/api/properties.js]: Maybe `to.be.an('object')` is a more precise assertion here.

**Comment by jpnelson** [comment]
@treshugart it seems like there's a problem with the saucelab build – I've updated the PR and the tests are now passing locally, but Chrome is failing to launch (on master too)

**Comment by treshugart** [comment]
@jpnelson I just restarted the builds. Definitely seems to be a Sauce Labs issue. We've been having some trouble with them.

File: test/unit/api/properties.js [hunk 8]
```diff
@@ -20,6 +20,70 @@ function testTypeValues (type, values) {
...
+      let fooCalled = false;
+      elem.addEventListener('propertychange', () => propertyChangeCalled = true);
+      elem.addEventListener('foo', () => fooCalled = true);
+      elem.test = true;
+
+      expect(propertyChangeCalled).to.equal(false);
+      expect(fooCalled).to.equal(true);
+    });
+
+    it('is triggered only when properties change', function () {
```
Comment (treshugart) [inline test/unit/api/properties.js]: Should probably also have a test to make sure it doesn't bubble.

File: src/lifecycle/properties-init.js [hunk 9]
```diff
@@ -150,6 +161,18 @@ function createNativePropertyDefinition (name, opts) {
       newValue = opts.coerce(newValue);
     }
 
+    const propertyHasChanged = newValue !== oldValue;
+    if (propertyHasChanged && opts.event) {
+      const cancelled = triggerPropertyChangedEvent(this, {
```
Comment (treshugart) [inline src/lifecycle/properties-init.js]: Is it worth having to call out to another function for this? I don't see it abstracting much.

**Comment by treshugart** [comment]
lgtm

**Comment by joscha** [comment]
LGTM - As a side note: I'd like to propose we replace the CustomEvent impl. with the one from AUI in the longterm.

**Pull request merged**